src/CodeReview.Orchestrator/Services/DockerContextSwitcher.cs

```diff
@@ -0,0 +1,162 @@
+            if (requirements == null)
+                throw new ArgumentNullException(nameof(requirements));
+
+            var matchingEngine = _engineProvider.Find(requirements);
+
+            if (_engineContext.Engine == null)
+            {
+                _engineContext.Engine = matchingEngine;
+                return;
+            }
+
+            if (_engineContext.Engine.Equals(matchingEngine))
+                return;
+            
+            await SwitchVolumesAsync(context, matchingEngine);
+        }
+
+        private async Task SwitchVolumesAsync(IProcessingContext context, DockerEngine newEngine)
+        {
+            using var folder = _tempFolderFactory.Create();
+            
+            var artifactsFolder = _pathService.Combine(folder.Path, "artifacts");
+            var srcFolder = _pathService.Combine(folder.Path, "src");
+            var importsFolder = _pathService.Combine(folder.Path, "imports");
+
+            var containerId = await CreateWorkerContainerAsync(context);
+
+            try
+            {
+                await ExportFolderContent(containerId, GetArtifactsFolderForOs(_engineContext.Engine.Os), artifactsFolder);
+                await ExportFolderContent(containerId, GetSrcFolderForOs(_engineContext.Engine.Os), srcFolder);
+                await ExportFolderContent(containerId, GetImportsFolderForOs(_engineContext.Engine.Os), importsFolder);
+            }
+            finally
+            {
+                await _containerService.RemoveContainerAsync(containerId);
+            }
+
+            await context.CleanUpVolumesAsync();
+
+            _engineContext.Engine = newEngine;
+
+            await context.InitializeAsync();
+
+            containerId = await CreateWorkerContainerAsync(context);
+
+            try
+            {
+                await ImportDataToContainerAsync(artifactsFolder, containerId, GetArtifactsFolderForOs(_engineContext.Engine.Os));
+                await ImportDataToContainerAsync(srcFolder, containerId, GetSrcFolderForOs(_engineContext.Engine.Os));
+                await ImportDataToContainerAsync(importsFolder, containerId, GetImportsFolderForOs(_engineContext.Engine.Os));
+            }
+            finally
+            {
+                await _containerService.RemoveContainerAsync(containerId);
+            }
+        }
+
+        private async Task ImportDataToContainerAsync(string dataFolder, string containerId, string containerFolder)
+        {
+            if (!_directoryService.Exists(dataFolder))
+                return;
+            
+            await using var outStream = _tarArchiveService.Create(dataFolder);
+
+            await _containerService.ImportFilesIntoContainerAsync(containerId, containerFolder, outStream);
+        }
+
+        private async Task<string> CreateWorkerContainerAsync(IProcessingContext context)
+        {
+            return await _containerService.CreateContainerAsync(
```

Comment:
Please, do not use async/await in such cases. Just return `Task<T>` to avoid using a state machine.